### PR TITLE
[WIP]vim-patch: 8.2.{1665, 1726} matchfuzzy()

### PR DESCRIPTION
**Pull request by vigoux**
Attempt (failed as of now) to port `matchfuzzy` and `matchfuzzypos`.
@janlazo may I beg for your help on this ?

#### vim-patch:8.2.1665: cannot do fuzzy string matching

Problem:    Cannot do fuzzy string matching.
Solution:   Add matchfuzzy(). (Yegappan Lakshmanan, closes vim/vim#6932)
vim/vim@635414d

#### vim-patch:8.2.1726: fuzzy matching only works on strings

Problem:    Fuzzy matching only works on strings.
Solution:   Support passing a dict.  Add matchfuzzypos() to also get the match
            positions. (Yegappan Lakshmanan, closes vim/vim#6947)
vim/vim@4f73b8e

### CODE REVIEW

File: src/nvim/search.c.rej [hunk 1]
```diff
@@ -0,0 +1,662 @@
+--- src/nvim/search.c
```
Comment (janlazo) [inline src/nvim/search.c.rej:1]: Remove this file.

Comment (vigoux) [reply src/nvim/search.c.rej]: File has been removed and lint error fixed. The tests are not passing for some reason though. Seems related to `TV_LIST_ITER, that I don't really know how to use.

File: src/nvim/search.c [hunk 2]
```diff
@@ -5061,3 +5061,531 @@ bool search_was_last_used(void)
...
+    return;
+  }
+
+  ptrs = xcalloc(sizeof(fuzzyItem_T), len);
+  if (ptrs == NULL) {
+    return;
+  }
+
+  // For all the string items in items, get the fuzzy matching score
+  TV_LIST_ITER(items, li, {
```
Comment (vigoux) [inline src/nvim/search.c:5389]: This does not seem to work and I really don't know why.

**Comment by janlazo** [comment]
@vigoux I'll get back to this PR after checking if there are older patches missing.

File: src/nvim/search.c (hunk 2 above)
Comment (janlazo) [reply src/nvim/search.c:5389]: Did you check https://github.com/neovim/neovim/wiki/List-management-in-Neovim?

File: src/nvim/testdir/test_matchfuzzy.vim [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+        \ ['xŗŝţũŵż', 'ŗŝţũŵż']->matchfuzzy('ŗŝţũŵż'))
+  " preference for sequential match
+  call assert_equal(['ㄞㄡㄤﬀﬁﬂ', 'ㄞaㄡbㄤcﬀdﬁeﬂ'],
+        \ ['ㄞaㄡbㄤcﬀdﬁeﬂ', 'ㄞㄡㄤﬀﬁﬂ']->matchfuzzy('ㄞㄡㄤﬀﬁﬂ'))
+  " non-matching leading letter(s) penalty
+  call assert_equal(['xㄞㄡㄤﬀﬁﬂ', 'xxㄞㄡㄤﬀﬁﬂ'],
+        \ ['xxㄞㄡㄤﬀﬁﬂ', 'xㄞㄡㄤﬀﬁﬂ']->matchfuzzy('ㄞㄡㄤﬀﬁﬂ'))
+  " total non-matching letter(s) penalty
+  call assert_equal(['ŗŝţ', 'ŗŝţx', 'ŗŝţxx'],
+        \ ['ŗŝţxx', 'ŗŝţ', 'ŗŝţx']->matchfuzzy('ŗŝţ'))
```
Comment (janlazo) [inline src/nvim/testdir/test_matchfuzzy.vim:150]: Method patches are not ported yet and they're difficult to port. Method refers to the syntax `(expr)->method(foo, bar)`  which is syntax sugar for `method((expr), foo, bar)`.

One could comment this out with a `TODO` comment but devs who merge the method patches may forgot to uncomment this.

Comment (vigoux) [reply src/nvim/testdir/test_matchfuzzy.vim:150]: I missed this one, I'll fix that soon.

File: src/nvim/search.c (hunk 2 above)
Comment (vigoux) [reply src/nvim/search.c:5389]: Yes I used it to port this patch.

Comment (vigoux) [reply src/nvim/search.c:5389]: @janlazo coming to this after a while, it is still a complete mistery  for me why this loop is not running at all.

**Comment by romgrk** [comment]
You might want to include this one at the same time:

**patch 8.2.1872: matchfuzzy() does not prefer sequential matches**

Problem:    Matchfuzzy() does not prefer sequential matches.
Solution:   Give sequential matches a higher bonus. (Christian Brabandt, closes https://github.com/vim/vim/pull/7140)
https://github.com/vim/vim/commit/e9f9f16387554e5b34ba42ce00c42c28dd66af58

**Comment by blankname** [comment]
Related:

**patch 8.2.1893: fuzzy matching does not support multiple words**

Problem:    Fuzzy matching does not support multiple words.
Solution:   Add support for matching white space separated words. (Yegappan
            Lakshmanan, closes vim/vim#7163)
vim/vim@8ded5b6

**patch 8.2.1921: fuzzy matching does not recognize path separators**

Problem:    Fuzzy matching does not recognize path separators.
Solution:   Add a bonus for slash and backslash. (Yegappan Lakshmanan,
            closes vim/vim#7225)
vim/vim@dcdd42a

**Comment by vigoux** [comment]
Yes I am adding those as I go.

File: src/nvim/search.c [hunk 4]
```diff
@@ -5106,3 +5106,527 @@ bool search_was_last_used(void)
...
+    l = tv_list_alloc(kListLenShouldKnow);
+    if (l == NULL) {
+      goto done;
+    }
+
+    tv_list_append_list(rettv->vval.v_list, l);
+  }
+
+  char buf[NUMBUFLEN];
+  if (tv_get_string_buf_chk(&argvars[1], buf)) {
```
Comment (romgrk) [inline src/nvim/search.c:5625]: Shouldn't you be using the return value of `tv_get_string_buf_chk()`? It uses `buf` *only* if `&argvars[1]` is not a string, IIUC.

File: src/nvim/search.c [hunk 5]
```diff
@@ -5106,3 +5106,527 @@ bool search_was_last_used(void)
...
+    typval_T rettv;
+    typval_T *tv_li = TV_LIST_ITEM_TV(li);
+
+    ptrs[i].item = li;
+    ptrs[i].score = SCORE_NONE;
+    itemstr = NULL;
+    rettv.v_type = VAR_UNKNOWN;
+
+    if (tv_check_str(tv_li)) {  // list of strings
+      itemstr = (char *)tv_li->vval.v_string;
```
Comment (romgrk) [inline src/nvim/search.c:5447]: Here `tv_check_str()` checks that a value can be *cast* to a string, not that it *is* a string.

File: src/nvim/search.c (hunk 4 above)
Comment (vigoux) [reply src/nvim/search.c:5625]: I don't really know tbh, I actually never ported patches from upstream vim

File: src/nvim/search.c (hunk 5 above)
Comment (vigoux) [reply src/nvim/search.c:5447]: Oh interesting

File: src/nvim/search.c (hunk 4 above)
Comment (romgrk) [reply src/nvim/search.c:5625]: Source says that you should must use the return value:

https://github.com/neovim/neovim/blob/4e6f00dd29332ce549006e8df1b1392ed4209954/src/nvim/eval/typval.c#L2967-L2980

File: src/nvim/search.c (hunk 2 above)
Comment (romgrk) [reply src/nvim/search.c:5435]: Do you have a test-case where it's not running? I've ran the PR and with the 2 other comments it seemed to run fine for simple cases.

**Comment by seandewar** [comment]
Continued in #16873 :+1:

**Pull request closed without merging**